### ISSUE TITLE
Consider adding initialization callback to runApp

### ISSUE THREAD

**Issue by matthew-carroll** [flutter/flutter#64288]
Developers often need to run some set of behaviors near app initialization that require various Flutter services. The typical remedy is to invoke some kind of direct initialization of an internal Flutter binding before calling `runApp()`, for example:

```dart
void main() {
  WidgetsFlutterBinding.ensureInitialized();

  // Run some Flutter-relevant code, like early plugin communication

  runApp(MyApp());
}
```

Consider offering an optional callback in `runApp()` that is invoked after all binding initializations so that Flutter developers do not need to understand the implementation details of `runApp()` such that those developers manually orchestrate the internal initialization:

```dart
void main() {
  runApp(
    MyApp,
    onFlutterInitialized: () {
      // Run some Flutter-relevant code, like early plugin communication
    },
  );
}
```

**Comment by frank06** [comment]
How about running a loader app while waiting for initialization?

```dart
void main() async {
  runApp(LoaderApp());

  // Run some Flutter-relevant code, like early plugin communication

  runApp(MyApp());
}
```

**Comment by matthew-carroll** [comment]
@frank06 can you elaborate on why that would be a preferable approach? To me it appears to be more confusing, far more difficult to implement, and would likely generate as much confusion as the existing approach.

**Comment by matthew-carroll** [comment]
Interesting. It sounds like that would expose the bindings and therefore make it obvious that they're being initialized. Seems like a possible solution to the copy/paste issue by ensuring that the relevant lines are always present.

Would this make "hello world" projects more daunting?

**Comment by chunhtai** [comment]
I am a bit opposing the idea of separating out the binding initializing from the runApp. This will add more boilerplate code if we do this, and it will be confusing to beginner anyway if they don't care about the binding.

 I feel we will end up asking user to not use runApp but do bootstrapping all by themselves if we keep going down this road.

**Comment by Hixie** [comment]
Yeah. It's not a great option. Unless you need to deal with the binding, it's nice that it's just taken core of for you.

I'm not a fan of APIs that do stuff for you that you can basically do yourself, though. I think maybe the current situation is the right API and we need to work on our documentation and/or error messages if people are still tripping on it.

**Comment by derolf** [comment]
TBH: every serious Flutter project will initialize bindings before runApp to setup some stuff. I would hence remove it from runApp.

**Comment by untp** [comment]
There is another alternative, extending `WidgetsFlutterBinding`, e.g.

```dart
void main() => runZonedGuarded(() => MainBinding(), ErrorHandler.onError);

class MainBinding extends WidgetsFlutterBinding {
  MainBinding() {
    build().then(runApp);
  }

  Future<Widget> build() async {
    FlutterError.onError = ErrorHandler.onFlutterError;
    await Firebase.initializeApp();
    // await other initializations
    return MultiProvider(
      providers: [
        Provider.value(value: await SharedPreferences.getInstance()),
        // add other providers
      ],
      child: const App(),
    );
  }
}

```

I use this on my projects, it looks better for me.
